internal/cache/resource_map.go

```diff
@@ -1,153 +1,126 @@
-	if e.isDeleted {
-		// In between loading the entry and acquiring the read lock, the entry was deleted, do nothing.
-		return false
-	}
+func (m *ResourceMap[K, V]) withLock(key K, f func()) {
+	lock := &m.locks[maphash.Comparable(seed, key)%m.stripes]
+	lock.Lock()
+	defer lock.Unlock()
 
-	compute(key, e.value)
-	return true
+	f()
 }
 
-// Compute ensures that an entry for the given name exists in the map before executing the given compute function. If
-// the entry was already in the map, it has the same semantics as ComputeIfPresent. Otherwise, it uses the given
-// newValue constructor to create a new value before executing the given compute method, and no ComputeIfPresent will
-// run until both the newValue constructor and the compute function complete. This is to prevent
-// subsequent ComputeIfPresent operations from reading a partially initialized value.
-func (m *ResourceMap[K, T]) Compute(
+// Compute first creates the value for the given key using the given function if no corresponding
+// entry exists. Then it executes the given compute function
+func (m *ResourceMap[K, V]) Compute(
 	key K,
-	newValue func(key K) T,
-	compute func(key K, value T),
-) {
-	// Do an initial check to see if the entry is already present in the map, and if so, apply the function.
-	if m.ComputeIfPresent(key, compute) {
-		return
-	}
-
-	// Otherwise, attempt to create the entry
-	e := new(resourceMapEntry[T])
-	// Inserting the entry while already holding the write lock means another routine that loads this entry from the
-	// map will be forced to wait until it is fully initialized before reading its value.
-	e.lock.Lock()
-	defer e.lock.Unlock()
-
-	// This loop guarantees the creation of the entry by repeatedly attempting to insert the value in the map,
-	// guaranteeing that the entry is eventually inserted.
-	for {
-		// Attempt to insert the new resourceMapEntry in the map
-		eAny, didNotStore := m.syncMap.LoadOrStore(key, e)
-		if !didNotStore {
-			// Entry successfully inserted in the map, the rest of the function can initialize it safely as it is
-			// currently holding the write lock.
-			break
+	newValue func(key K) V,
+	compute func(value V),
+) (v V) {
+	m.withLock(key, func() {
+		var ok bool
+		v, ok = m.load(key)
+		if !ok {
+			v = newValue(key)
+			m.resources.Store(key, v)
+			m.size.Add(1)
 		}
 
-		// Unhappy path: between the original ComputeIfPresent check and now, another routine created the entry.
-		e := eAny.(*resourceMapEntry[T])
-		e.lock.RLock()
-		if !e.isDeleted {
-			// The entry was already present in the map, execute the compute function safely since the read lock is
-			// currently held.
-			compute(key, e.value)
-			// Nothing left to be done, exit the function
-			e.lock.RUnlock()
+		compute(v)
+	})
+
+	return v
+}
+
+// ComputeIfPresent invokes the given function only if a corresponding entry exists in the map for
+// the given key.
+func (m *ResourceMap[K, V]) ComputeIfPresent(key K, compute func(value V)) {
```

Comment:
consider to return the value computed by the function or null if the map is not changed?

Reply:
The signature of the function actually prevents it from being modified! `compute` has no return value, so this only executes on the value without updating it.

Reply:
I can return a bool if the function was executed though.